core/computed/entity-classification.js

```diff
@@ -0,0 +1,110 @@
+import {Util} from '../util.cjs';
+import UrlUtils from '../lib/url-utils.js';
+import thirdPartyWeb from '../lib/third-party-web.js';
+
+/** @typedef {Map<string, LH.Artifacts.Entity>} EntityCache */
+
+class EntityClassification {
+  /**
+   * @param {EntityCache} entityCache
+   * @param {string} url
+   * @return Entity | undefined
+   */
+  static makeUpAnEntity(entityCache, url) {
+    // We can make up an entity only for those URLs with a valid domain attached.
+    // So we further restrict from allowed URLs to (/^h/ or http[s] and h2).
+    if (!(UrlUtils.isValid(url) && UrlUtils.isProtocolAllowed(url) &&
+      Util.createOrReturnURL(url).protocol[0] === 'h')) return;
+    const rootDomain = Util.getRootDomain(url);
+    if (!rootDomain) return;
+    if (entityCache.has(rootDomain)) return entityCache.get(rootDomain);
+
+    const unrecognizedEntity = {
+      name: rootDomain,
+      company: rootDomain,
+      category: '',
+      categories: [],
+      domains: [rootDomain],
+      averageExecutionTime: 0,
+      totalExecutionTime: 0,
+      totalOccurrences: 0,
+      isUnrecognized: true,
+    };
+    entityCache.set(rootDomain, unrecognizedEntity);
+    return unrecognizedEntity;
+  }
+
+  /**
+   * @param {{URL: LH.Artifacts['URL'], devtoolsLog: LH.DevtoolsLog}} data
+   * @param {LH.Artifacts.ComputedContext} context
+   * @return {Promise<LH.Artifacts.EntityClassification>}
+   */
+  static async compute_(data, context) {
+    const networkRecords = await NetworkRecords.request(data.devtoolsLog, context);
+    /** @type {EntityCache} */
+    const madeUpEntityCache = new Map();
+    /** @type {Map<string, LH.Artifacts.Entity>} */
+    const urlToEntity = new Map();
+    /** @type {Map<LH.Artifacts.Entity, Array<string>>} */
+    const entityToURLs = new Map();
+
+    for (const record of networkRecords) {
+      const {url} = record;
+      if (urlToEntity.has(url)) continue;
+
+      const entity = thirdPartyWeb.getEntity(url) ||
+        EntityClassification.makeUpAnEntity(madeUpEntityCache, url);
+      if (!entity) continue;
+
+      const entityURLs = entityToURLs.get(entity) || [];
+      entityURLs.push(url);
+      entityToURLs.set(entity, entityURLs);
+      urlToEntity.set(url, entity);
+    }
+
+    // When available, first party identification will be done via
+    // `mainDocumentUrl` (for navigations), and falls back to `finalDisplayedUrl` (for timespan/snapshot).
+    // See https://github.com/GoogleChrome/lighthouse/issues/13706
+    /** @type {LH.Artifacts.Entity | undefined} */
+    let firstParty;
+    const firstPartyUrl = data.URL?.mainDocumentUrl || data.URL?.finalDisplayedUrl;
+    if (firstPartyUrl) {
+      firstParty = thirdPartyWeb.getEntity(firstPartyUrl) ||
+        EntityClassification.makeUpAnEntity(madeUpEntityCache, firstPartyUrl);
+    }
+
+    /**
+     * Convenience function to return the entity name for a URL.
+     * @param {string} url
+     * @returns string | undefined
```

Comment:
```suggestion
     * @return {string | undefined}
```

Reply:
Will fix the `{}`. Curious why we prefer `@return` over the actual JSDoc tag? (`@returns` is the recommended tag, and `@return` is a synonym?).

Reply:
I don't think there is a good reason we picked one over the other, but at this point we should be consistent.

Reply:
Yeah that makes sense. I was wondering if there was a reason beyond the obvious :)